libsolidity/experimental/ast/TypeSystem.cpp

```diff
@@ -0,0 +1,351 @@
+	return {};
+}
+
+experimental::Type TypeEnvironment::resolve(Type _type) const
+{
+	Type result = _type;
+	while (auto const* var = std::get_if<TypeVariable>(&result))
+		if (Type const* resolvedType = util::valueOrNullptr(m_typeVariables, var->index()))
+			result = *resolvedType;
+		else
+			break;
+	return result;
+}
+
+experimental::Type TypeEnvironment::resolveRecursive(Type _type) const
+{
+	return std::visit(util::GenericVisitor{
+		[&](TypeConstant const& _type) -> Type {
+			return TypeConstant{
+				_type.constructor,
+				_type.arguments | ranges::views::transform([&](Type _argType) {
+					return resolveRecursive(_argType);
+				}) | ranges::to<std::vector<Type>>
+			};
+		},
+		[&](TypeVariable const&) -> Type {
+			return _type;
+		},
+		[&](std::monostate) -> Type {
+			return _type;
+		}
+	}, resolve(_type));
+}
+
+Sort TypeEnvironment::sort(Type _type) const
+{
+	return std::visit(util::GenericVisitor{
+		[&](TypeConstant const& _expression) -> Sort
+		{
+			auto const& constructorInfo = m_typeSystem.constructorInfo(_expression.constructor);
+			auto argumentSorts = _expression.arguments | ranges::views::transform([&](Type _argumentType) {
+				return sort(resolve(_argumentType));
+			}) | ranges::to<std::vector<Sort>>;
+			Sort sort;
+			for (auto const& arity: constructorInfo.arities)
+			{
+				solAssert(arity.argumentSorts.size() == argumentSorts.size());
+				bool hasArity = true;
+				for (auto&& [argumentSort, arityArgumentSort]: ranges::zip_view(argumentSorts, arity.argumentSorts))
+				{
+					if (!(arityArgumentSort <= argumentSort))
+					{
+						hasArity = false;
+						break;
+					}
+				}
+
+				if (hasArity)
+					sort.classes.insert(arity.typeClass);
+			}
+			return sort;
+		},
+		[](TypeVariable const& _variable) -> Sort { return _variable.sort(); },
+		[](std::monostate) -> Sort { solAssert(false); }
+	}, _type);
+}
+
+TypeConstructor TypeSystem::declareTypeConstructor(std::string _name, std::string _canonicalName, size_t _arguments, Declaration const* _declaration)
+{
+	solAssert(m_canonicalTypeNames.insert(_canonicalName).second, "Duplicate canonical type name.");
+	Sort baseSort{{primitiveClass(PrimitiveClass::Type)}};
+	size_t index = m_typeConstructors.size();
+	m_typeConstructors.emplace_back(TypeConstructorInfo{
+		_name,
+		_canonicalName,
+		{Arity{std::vector<Sort>{_arguments, baseSort}, primitiveClass(PrimitiveClass::Type)}},
+		_declaration
+	});
+	TypeConstructor constructor{index};
+	if (_arguments)
+	{
+		std::vector<Sort> argumentSorts;
+		std::generate_n(std::back_inserter(argumentSorts), _arguments, [&](){ return Sort{{primitiveClass(PrimitiveClass::Type)}}; });
+		std::vector<Type> argumentTypes;
+		std::generate_n(std::back_inserter(argumentTypes), _arguments, [&](){ return freshVariable({}); });
+		auto error = instantiateClass(type(constructor, argumentTypes), Arity{argumentSorts, primitiveClass(PrimitiveClass::Type)});
+		solAssert(!error, *error);
+	}
```

Comment:
I don't get the purpose of `instantiateClass()` call here. Unless I'm missing something, it will simply insert a second, identical copy of `PrimitiveClass::Type` arity into `arities` and with how it's called, validations inside can never fail (we're calling it on a type constant and using matching numbers of args).

Is this bit just redundant? If so, why is it done only when there is at least one argument?

Reply:
The system currently has a primitive class ``Type`` that contains all types (ultimately, we may not need that, actually, but currently at least it's there). When you declare a new type constructor, you need to clarify that it, applied to arguments of ``Type`` sorts gives you something of class ``Type``.

Reply:
But why do we need it two identical copies of that class in `arities`?

Reply:
Ah, yeah, we don't :-) - we should probably not add it manually in here and only let ``instantiateClass`` handle it, I need to read through it properly myself again :-).